distribution/tools/keystore-cli/src/main/java/org/elasticsearch/common/settings/BaseKeyStoreCommand.java

```diff
@@ -0,0 +1,107 @@
+                } else {
+                    if (terminal.promptYesNo("The elasticsearch keystore does not exist. Do you want to create it?", false) == false) {
+                        terminal.println("Exiting without creating keystore.");
+                        return;
+                    }
+                }
+                // allow the user to decide if an auto-created keystore should be password protected
+                keyStorePassword =
+                    terminal.promptYesNo("Do you want to protect the keystore with a password?", false) == false ?
+                        new SecureString(new char[0]) : readPassword(terminal, true);
+                keyStore = KeyStoreWrapper.create();
+                keyStore.save(configFile, keyStorePassword.getChars());
+            } else {
+                keyStorePassword = keyStore.hasPassword() ? readPassword(terminal, false) : new SecureString(new char[0]);
+                keyStore.decrypt(keyStorePassword.getChars());
+            }
+            executeCommand(terminal, options, env);
+        } catch (SecurityException e) {
+            throw new UserException(ExitCodes.DATA_ERROR, e.getMessage());
+        } finally {
+            clearPassword();
+        }
+    }
+
+    private void clearPassword() {
+        if (keyStorePassword != null) {
+            keyStorePassword.close();
+        }
+    }
+
+    /**
+     * Reads the keystore password from the {@link Terminal}, prompting for verification where applicable and returns it as a
+     * {@link SecureString}.
+     *
+     * @param terminal         the terminal to use for user inputs
+     * @param withVerification whether the user should be prompted for password verification
+     * @return a SecureString with the password the user entered
+     * @throws UserException If the user is prompted for verification and enters a different password
+     */
+    static SecureString readPassword(Terminal terminal, boolean withVerification) throws UserException {
+        final char[] passwordArray;
+        if (withVerification) {
+            passwordArray = terminal.readSecret("Enter new password for the elasticsearch keystore (empty for no password): ");
+            char[] passwordVerification = terminal.readSecret("Enter same password again: ");
+            if (Arrays.equals(passwordArray, passwordVerification) == false) {
+                throw new UserException(ExitCodes.DATA_ERROR, "Passwords are not equal, exiting.");
+            }
+            Arrays.fill(passwordVerification, '\u0000');
+        } else {
+            passwordArray = terminal.readSecret("Enter password for the elasticsearch keystore : ");
+        }
+        final SecureString password = new SecureString(passwordArray);
+        return password;
+    }
+
+    protected abstract void executeCommand(Terminal terminal, OptionSet options, Environment env) throws Exception;
```

Comment:
please add a basic javadoc explaning when this is called, the keystore and keystore password are available.